### PR TITLE
feat: add hash-checkpoint contract

### PR DESCRIPTION
**Pull request by DavidMinarsch**

### CODE REVIEW

File: contracts/utils/HashCheckpoint.sol [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    error ZeroValue();
+    event OwnerUpdated(address indexed owner);
+    event ManagerUpdated(address indexed manager);
+    event BaseURIChanged(string baseURI);
+    event HashUpdated(address indexed emitter, bytes32 hash);
+
+    // Owner address
+    address public owner;
+    // Unit manager
+    address public manager;
```
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply contracts/utils/HashCheckpoint.sol:18]: No - removed now.

File: contracts/utils/HashCheckpoint.sol [hunk 2]
```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.21;
+
+/// @title Generic Registry - Smart contract for generic registry template
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+abstract contract HashCheckpoint {
+    error OwnerOnly(address sender, address owner);
```
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply contracts/utils/HashCheckpoint.sol:7]: Thanks

File: contracts/utils/HashCheckpoint.sol [hunk 3]
```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.21;
+
+/// @title Hash Checkpoint - Smart contract for checkpointing IPFS hashes on-chain
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
```
Comment (kupermind) [inline contracts/utils/HashCheckpoint.sol:5]: We need to change the author :D

File: contracts/utils/HashCheckpoint.sol [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    /// @dev Hash checkpoint constructor.
+    /// @param _baseURI Hash registry base URI.
+    constructor(string memory _baseURI) {
+        baseURI = _baseURI;
+        owner = msg.sender;
+    }
+
+    /// @dev Changes the owner address.
+    /// @param newOwner Address of a new owner.
+    function changeOwner(address newOwner) external virtual {
```
Comment (kupermind) [inline contracts/utils/HashCheckpoint.sol:41]: Probably no need to be a virtual function? For all of them below as well.

File: contracts/utils/HashCheckpoint.sol (hunk 3 above)
Comment (DavidMinarsch) [reply contracts/utils/HashCheckpoint.sol:5]: Lol you'll be the author eventually

File: contracts/utils/HashCheckpoint.sol (hunk 4 above)
Comment (DavidMinarsch) [reply contracts/utils/HashCheckpoint.sol:41]: I guess no harm

**Review by kupermind — APPROVED** [review]

**Pull request merged**